public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -70,193 +91,297 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
+      catchError((err: any) => this.throwUnless(err, err.cancelled || err.status === 404, target)),
       filter((response: any) => (response.cancelled ? false : true)),
-      map((response: any) => {
-        const data = this.processResult(response.data, query);
-        return { data, key: query.refId };
-      })
+      switchMap((response: { data: LokiResponse; status: number }) =>
+        iif<DataQueryResponse, DataQueryResponse>(
+          () => response.status === 404,
+          defer(() => this.runLegacyQuery(target, options)),
+          defer(() =>
+            processRangeQueryResponse(response.data, target, query, responseListLength, this.maxLines, options.reverse)
+          )
+        )
+      )
     );
   };
 
-  query(options: DataQueryRequest<LokiQuery>): Observable<DataQueryResponse> {
-    const subQueries = options.targets
-      .filter(target => target.expr && !target.hide)
-      .map(target => {
-        if (target.liveStreaming) {
-          return this.runLiveQuery(options, target);
-        }
-        return this.runQuery(options, target);
-      });
+  createLegacyLiveTarget(target: LokiQuery, options: { maxDataPoints?: number }): LegacyTarget {
+    const { query, regexp } = parseQuery(target.expr);
+    const baseUrl = this.instanceSettings.url;
+    const params = serializeParams({ query });
 
-    // No valid targets, return the empty result to save a round trip.
-    if (isEmpty(subQueries)) {
-      return of({
-        data: [],
-        state: LoadingState.Done,
-      });
-    }
+    return {
+      query,
+      regexp,
+      url: convertToWebSocketUrl(`${baseUrl}/api/prom/tail?${params}`),
+      refId: target.refId,
+      size: Math.min(options.maxDataPoints || Infinity, this.maxLines),
+    };
+  }
 
-    return merge(...subQueries);
+  createLiveTarget(target: LokiQuery, options: { maxDataPoints?: number }): LegacyTarget {
+    const { query, regexp } = parseQuery(target.expr);
+    const baseUrl = this.instanceSettings.url;
+    const params = serializeParams({ query });
+
+    return {
+      query,
+      regexp,
+      url: convertToWebSocketUrl(`${baseUrl}/loki/api/v1/tail?${params}`),
+      refId: target.refId,
+      size: Math.min(options.maxDataPoints || Infinity, this.maxLines),
+    };
   }
 
+  /**
+   * Runs live queries which in this case means creating a websocket and listening on it for new logs.
+   * This returns a bit different dataFrame than runQueries as it returns single dataframe even if there are multiple
+   * Loki streams, sets only common labels on dataframe.labels and has additional dataframe.fields.labels for unique
+   * labels per row.
+   */
+  runLiveQuery = (target: LokiQuery, options: { maxDataPoints?: number }): Observable<DataQueryResponse> => {
+    const liveTarget = this.createLiveTarget(target, options);
+    const stream = this.streams.getStream(liveTarget);
+
+    return stream.pipe(
+      catchError((err: any) => {
+        if (err.status === 404) {
```

Comment:
AFAIK Websocket connection errors don't have `status` codes, so this will not recover to the legacy API.